### ISSUE TITLE
ECMAScript standard is misinterpreted with respect to non-writable properties

### ISSUE THREAD

**Issue by acarstoiu** [joyent/node#25620]
This issue was discovered by a JavaScript-learning novice who "didn't really know what he was doing" (his own words). It seems that the buggy ECMAScript test suite has managed to trick all major interpreter producers into thinking that it's all right to forbid creating a new property on an object if one of its prototypes already has such a property, with the _writable_ attribute set to `false`.

Here is the code that builds two objects, an `obj` having as prototype `proto`, both with two identically named properties (`name` and `surname`):
``` javascript
(function () {
'use strict';

var proto = Object.create(null, {
    name: {
        writable: false,
        enumerable: false,
        configurable: true,
        value: "name in proto"
    },
    surname: {
        writable: true,
        enumerable: false,
        configurable: false,
        value: "surname in proto"
    }
});

var obj = Object.create(proto);
console.log(Object.getPrototypeOf(obj) === proto);

// plain assignment, gets rejected!
obj.name = "name in obj";
// elaborate assignment, passes through
/* Object.defineProperty(obj, "name", {
    writable: false,
    enumerable: false,
    configurable: true,
    value: "name in obj"
}); */

// plain assignment, is accepted!
obj.surname = "surname in obj";

console.log(obj, obj.name);
console.log(proto, proto.name, proto.surname);
}) ();
```
The snippet proves the following:
- a property assigned to `obj` belongs to it, **not to its prototype**
- a direct assignment of a property is subject to writability verification using the metadata found in a _prototype_

This is wrong, no matter how you look at it.
Writing access to a property should be governed by the property attributes of the destination object, because _that is what is affected_. It is only the _reading access_, which searches in the prototype chain for the addressed property, as a fallback, that should indeed be governed by the prototype metadata.
Even if we accept that this "feature" has to do with security, it actually does not protect anybody from anything. Just uncomment the elaborate assignment that uses `Object.defineProperty()` and comment the simple assignment of `obj.name`, to prove it.

Node.js 0.10.39 does not even bother to throw an error, while a newer V8 (tested with Chrome 43.0) does throw. Also, all other major browsers throw an error stating that I'm not allowed to change a read-only property, except that **I was not**: IE 11, Safari 8.0 and Firefox 38.0.

**Comment by cjihrig** [comment]
I don't think there is anything to be done in Node. V8 would be the right place to take this issue.

**Comment by acarstoiu** [comment]
I posted the issue here since I am under the impression that you people are patching V8 a bit before packaging it with Node.js.

In the meantime, I found the reason  - it's political, the specification was made that way for "backwards compatibility", but actually it's a **mistake** that they lost the chance to correct in ES5 (see [this](http://www.ecma-international.org/ecma-262/5.1/index.html#sec-8.12.4)) and ES6 (see [this](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-ordinary-object-internal-methods-and-internal-slots-set-p-v-receiver)). The specification lead's answer in this [particular discussion](http://wiki.ecmascript.org/doku.php?id=strawman:fixing_override_mistake) shows the clearly that there are no logical grounds for this nonsense.
